activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -107,7 +107,12 @@ def select_all(arel, name = nil, binds = [], preparable: nil, async: false) # :n
 
           if async
             result = lookup_sql_cache(sql, name, binds) || super(sql, name, binds, preparable: preparable, async: async)
-            FutureResult::Complete.new(result)
+
+            if result.is_a?(FutureResult)
```

Comment:
I tried your fix still same issue.
I found with that change it worked.

```suggestion
            if result.is_a?(FutureResult::Complete)
```

```ruby
result.class
ActiveRecord::FutureResult::Complete

result
#<ActiveRecord::FutureResult::Complete:0x0000000142b451c0
 @result=
  #<ActiveRecord::Result:0x0000000141c96778
   @column_types=
    {0=>#<ActiveModel::Type::Integer:0x0000000142271508 @limit=8, @precision=nil, @range=-9223372036854775808...9223372036854775808, @scale=nil>,
     "count"=>#<ActiveModel::Type::Integer:0x0000000142271508 @limit=8, @precision=nil, @range=-9223372036854775808...9223372036854775808, @scale=nil>},
   @columns=["count"],
   @hash_rows=nil,
   @rows=[[70]]>>

(ruby@puma: cluster worker 3: 94614 [yespark-rails]#95708) result.is_a?(FutureResult)
false
(ruby@puma: cluster worker 3: 94614 [yespark-rails]#95708) result.is_a?(FutureResult::Complete)
true
```

Reply:
You can find a reproducible case there:
https://gist.github.com/navidemad/e4af727d335faa3148d4675e698cba4e

Reply:
Can you test now?

Reply:
sorry for the delay, it works with your latest change